api/src/zenbox_smt/jni_calls.rs

```diff
@@ -0,0 +1,308 @@
+use jni::JNIEnv;
+use jni::objects::{JClass, JString, JObject, JValue};
+use jni::sys::{jbyteArray, jboolean, jint, jlong, jlongArray, jobject, jobjectArray};
+use crate::{read_raw_pointer, read_mut_raw_pointer};
+use crate::ginger_calls::{FieldElement, leaf_to_index, GingerMHTPath, Error};
+use crate::zenbox_smt::*;
+use std::any::TypeId;
+use std::ptr::null_mut;
+use jni::strings::JavaStr;
+
+fn get_field_name<'a, T: 'static>() -> &'a str {
+    if TypeId::of::<T>() == TypeId::of::<ZenBoxSMT>(){
+        "merkleTreePointer"
+    } else if TypeId::of::<T>() == TypeId::of::<FieldElement>(){
+        "fieldElementPointer"
+    } else {
+        panic!("Unknown type of a pointer")
+    }
+}
+
+fn get_raw_ptr<T: 'static>(env: &JNIEnv, ptr: JObject) -> *mut T {
+    let field_name = get_field_name::<T>();
+    let fe = env.get_field(ptr, field_name, "J")
+        .expect(&("Should be able to get field ".to_owned() + field_name));
+    fe.j().unwrap() as *mut T
+}
+
+fn unwrap_ptr<'a, T: 'static>(env: &JNIEnv, ptr: JObject) -> &'a T {
+    read_raw_pointer(get_raw_ptr(env, ptr))
+}
+
+fn unwrap_mut_ptr<'a, T: 'static>(env: &JNIEnv, ptr: JObject) -> &'a mut T {
+    read_mut_raw_pointer(get_raw_ptr(env, ptr))
+}
+
```

Comment:
Nice way to generalize, and it will be good if you apply it to all the JNI functions we have.
Of course this goes outside the scope of your task, but there's an issue opened about this:

https://github.com/HorizenOfficial/zendoo-sc-cryptolib/issues/22

We will do this kind of refactor, along with other refactoring and optimizations, when someone will deal with this issue. My suggestion is, for the moment, remove the functions above and use the ones already defined in api/lib.rs (just for the sake of uniformity), but save them somewhere for later

Reply:
Thanks Daniele! No problem, in my spare time I a can do at list partially such a refactoring by using these methods in lib.rs.